<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-preparation" data-toc-modified-id="Data-preparation-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Data preparation</a></span><ul class="toc-item"><li><span><a href="#We-import-the-modules-and-libraries-necessary-for-work" data-toc-modified-id="We-import-the-modules-and-libraries-necessary-for-work-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>We import the modules and libraries necessary for work</a></span></li><li><span><a href="#Let's-open-the-table" data-toc-modified-id="Let's-open-the-table-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Let's open the table</a></span></li><li><span><a href="#Let's-study-the-data" data-toc-modified-id="Let's-study-the-data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Let's study the data</a></span></li><li><span><a href="#Rename-columns" data-toc-modified-id="Rename-columns-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Rename columns</a></span></li><li><span><a href="#View-basic-information" data-toc-modified-id="View-basic-information-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>View basic information</a></span></li><li><span><a href="#Get-rid-of-null-values-in-the-&quot;tenure&quot;-column" data-toc-modified-id="Get-rid-of-null-values-in-the-&quot;tenure&quot;-column-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Get rid of null values in the "tenure" column</a></span></li><li><span><a href="#Let's-delete-the-column-with-the-row-index,-since-the-values-​​in-it-match-the-index,-and-the-columns-with-the-user's-last-name-and-id,-since-we-don't-need-them-for-further-analysis." data-toc-modified-id="Let's-delete-the-column-with-the-row-index,-since-the-values-​​in-it-match-the-index,-and-the-columns-with-the-user's-last-name-and-id,-since-we-don't-need-them-for-further-analysis.-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Let's delete the column with the row index, since the values ​​in it match the index, and the columns with the user's last name and id, since we don't need them for further analysis.</a></span></li><li><span><a href="#Using-OneHotEncoding" data-toc-modified-id="Using-OneHotEncoding-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Using OneHotEncoding</a></span></li><li><span><a href="#Divide-data-into-features-and-target-feature" data-toc-modified-id="Divide-data-into-features-and-target-feature-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Divide data into features and target feature</a></span></li><li><span><a href="#Let's-see-the-number-of-objects-of-negative-and-positive-class" data-toc-modified-id="Let's-see-the-number-of-objects-of-negative-and-positive-class-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Let's see the number of objects of negative and positive class</a></span></li><li><span><a href="#We-divide-the-initial-data-into-training,-validation-and-test-sets.-Let's-take-the-ratio-of-60/20/20%,-respectively.-To-begin-with,-let's-try-without-taking-into-account-the-imbalance" data-toc-modified-id="We-divide-the-initial-data-into-training,-validation-and-test-sets.-Let's-take-the-ratio-of-60/20/20%,-respectively.-To-begin-with,-let's-try-without-taking-into-account-the-imbalance-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>We divide the initial data into training, validation and test sets. Let's take the ratio of 60/20/20%, respectively. To begin with, let's try without taking into account the imbalance</a></span></li><li><span><a href="#Using-scaler" data-toc-modified-id="Using-scaler-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>Using scaler</a></span></li></ul></li><li><span><a href="#Problem-research" data-toc-modified-id="Problem-research-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Problem research</a></span><ul class="toc-item"><li><span><a href="#Using-a-decision-tree" data-toc-modified-id="Using-a-decision-tree-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Using a decision tree</a></span></li><li><span><a href="#Let's-try-random-forest" data-toc-modified-id="Let's-try-random-forest-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Let's try random forest</a></span></li><li><span><a href="#And-let's-see-the-logistic-regression" data-toc-modified-id="And-let's-see-the-logistic-regression-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>And let's see the logistic regression</a></span></li></ul></li><li><span><a href="#Fighting-imbalance" data-toc-modified-id="Fighting-imbalance-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Fighting imbalance</a></span><ul class="toc-item"><li><span><a href="#Upsampling" data-toc-modified-id="Upsampling-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Upsampling</a></span></li><li><span><a href="#Check-the-decision-tree-after-correcting-the-imbalance" data-toc-modified-id="Check-the-decision-tree-after-correcting-the-imbalance-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Check the decision tree after correcting the imbalance</a></span></li><li><span><a href="#Now-random-forest" data-toc-modified-id="Now-random-forest-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Now random forest</a></span></li><li><span><a href="#Downsampling" data-toc-modified-id="Downsampling-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Downsampling</a></span></li><li><span><a href="#Let's-check-f1-in-the-decision-tree" data-toc-modified-id="Let's-check-f1-in-the-decision-tree-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Let's check f1 in the decision tree</a></span></li><li><span><a href="#Now-let's-check-the-random-forest" data-toc-modified-id="Now-let's-check-the-random-forest-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Now let's check the random forest</a></span></li><li><span><a href="#And-check-the-logistic-regression" data-toc-modified-id="And-check-the-logistic-regression-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>And check the logistic regression</a></span></li></ul></li><li><span><a href="#Model-testing" data-toc-modified-id="Model-testing-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model testing</a></span><ul class="toc-item"><li><span><a href="#Conclusion." data-toc-modified-id="Conclusion.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Conclusion.</a></span></li></ul></li></ul></div>

In [1]:
!pip install scikit-learn==1.1.3

In [2]:
!pip install imblearn

# Customer churn

Customers began to leave Beta-Bank. Every month. A little, but noticeable. Banking marketers figured it was cheaper to keep current customers than to attract new ones.

It is necessary to predict whether the client will leave the bank in the near future or not. You are provided with historical data on customer behavior and termination of agreements with the bank.

Build a model with an extremely large *F1*-measure. To pass the project successfully, you need to bring the metric to 0.59. Check the *F1*-measure on the test set yourself.

Additionally measure *AUC-ROC*, compare its value with *F1*-measure.

Data source: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Data preparation

Project implementation plan:

Load and prepare data.

We investigate the balance of classes, train the model without taking into account the imbalance. Let us briefly describe the conclusions.

Let's improve the quality of the model, taking into account the imbalance of classes. We will train different models and find the best one. Let us briefly describe the conclusions.

Let's do the final test.

### We import the modules and libraries necessary for work

In [3]:
from joblib import dump
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

from sklearn.metrics import (
    accuracy_score,
    f1_score,
    roc_auc_score,
    precision_score,
    recall_score,
    confusion_matrix,
    roc_curve
)

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder

from sklearn.dummy import DummyClassifier

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.utils import shuffle

### Let's open the table

In [4]:
try:
    data = pd.read_csv('/datasets/Churn.csv', sep=',')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/Churn.csv',
                       sep=',')

### Let's study the data

In [5]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Signs

RowNumber - row index in data

CustomerId - unique customer identifier

Surname - surname

CreditScore - credit rating

Geography - country of residence

gender

Age - age

Tenure - how many years a person has been a bank client

Balance - account balance

NumOfProducts - the number of bank products used by the client

HasCrCard - the presence of a credit card

IsActiveMember - client activity

EstimatedSalary - estimated salary


Target feature

Exited - the fact that the client left

### Rename columns

In [6]:
data.columns = ['row_number', 'customer_id', 'surname', 'credit_score', 'geography', 'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card', 'is_active_member', 'estimated_salary', 'exited']

### View basic information

In [7]:
data.describe(include='all').T

count unique     top  freq           mean           std  \
row_number        10000.0    NaN     NaN   NaN         5000.5    2886.89568   
customer_id       10000.0    NaN     NaN   NaN  15690940.5694  71936.186123   
surname             10000   2932   Smith    32            NaN           NaN   
credit_score      10000.0    NaN     NaN   NaN       650.5288     96.653299   
geography           10000      3  France  5014            NaN           NaN   
gender              10000      2    Male  5457            NaN           NaN   
age               10000.0    NaN     NaN   NaN        38.9218     10.487806   
tenure             9091.0    NaN     NaN   NaN        4.99769      2.894723   
balance           10000.0    NaN     NaN   NaN   76485.889288  62397.405202   
num_of_products   10000.0    NaN     NaN   NaN         1.5302      0.581654   
has_cr_card       10000.0    NaN     NaN   NaN         0.7055       0.45584   
is_active_member  10000.0    NaN     NaN   NaN         0.5151      0.499797   
estimated_salary  10000.0    NaN     NaN   NaN  100090.239881  57510.492818   
exited            10000.0    NaN     NaN   NaN         0.2037      0.402769   

                         min          25%         50%          75%         max  
row_number               1.0      2500.75      5000.5      7500.25     10000.0  
customer_id       15565701.0  15628528.25  15690738.0  15753233.75  15815690.0  
surname                  NaN          NaN         NaN          NaN         NaN  
credit_score           350.0        584.0       652.0        718.0       850.0  
geography                NaN          NaN         NaN          NaN         NaN  
gender                   NaN          NaN         NaN          NaN         NaN  
age                     18.0         32.0        37.0         44.0        92.0  
tenure                   0.0          2.0         5.0          7.0        10.0  
balance                  0.0          0.0    97198.54    127644.24   250898.09  
num_of_products          1.0          1.0         1.0          2.0         4.0  
has_cr_card              0.0          0.0         1.0          1.0         1.0  
is_active_member         0.0          0.0         1.0          1.0         1.0  
estimated_salary       11.58     51002.11  100193.915  149388.2475   199992.48  
exited                   0.0          0.0         0.0          0.0         1.0

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [9]:
data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [10]:
data['tenure'].isna().sum()

909

### Get rid of null values in the "tenure" column

In [11]:
data['tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: tenure, dtype: int64

By what principle the numbers are distributed, we do not know, so we fill in the median value

In [12]:
data['tenure'] = data['tenure'].mask(data['tenure'].isna(),
                                     np.random.uniform(1, 10, size=len(data)))
data['tenure'].isna().sum()

0

### Let's delete the column with the row index, since the values ​​in it match the index, and the columns with the user's last name and id, since we don't need them for further analysis.

In [13]:
data.drop(["row_number", "surname", "customer_id"], axis=1, inplace=True)

In [14]:
data['exited'].mean()

0.2037

This means that 20% of customers leave the bank

In [15]:
data.head()

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    France  Female   42     2.0       0.00                1   
1           608     Spain  Female   41     1.0   83807.86                1   
2           502    France  Female   42     8.0  159660.80                3   
3           699    France  Female   39     1.0       0.00                2   
4           850     Spain  Female   43     2.0  125510.82                1   

   has_cr_card  is_active_member  estimated_salary  exited  
0            1                 1         101348.88       1  
1            0                 1         112542.58       0  
2            1                 0         113931.57       1  
3            0                 0          93826.63       0  
4            1                 1          79084.10       0

### Using OneHotEncoding

In [16]:
data_ohe=pd.get_dummies(data, drop_first=True)
data_ohe.head()

credit_score  age  tenure    balance  num_of_products  has_cr_card  \
0           619   42     2.0       0.00                1            1   
1           608   41     1.0   83807.86                1            0   
2           502   42     8.0  159660.80                3            1   
3           699   39     1.0       0.00                2            0   
4           850   43     2.0  125510.82                1            1   

   is_active_member  estimated_salary  exited  geography_Germany  \
0                 1         101348.88       1                  0   
1                 1         112542.58       0                  0   
2                 0         113931.57       1                  0   
3                 0          93826.63       0                  0   
4                 1          79084.10       0                  0   

   geography_Spain  gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

### Divide data into features and target feature

In [17]:
features = data_ohe.drop(['exited'], axis=1)
target = data_ohe['exited']

### Let's see the number of objects of negative and positive class

In [18]:
features[target == 0].shape, features[target == 1].shape

((7963, 11), (2037, 11))

Как видим, имеется дисбаланс

### We divide the initial data into training, validation and test sets. Let's take the ratio of 60/20/20%, respectively. To begin with, let's try without taking into account the imbalance

In [19]:
features_train, features_valid, target_train, target_valid = (
    train_test_split(
        features, target, test_size=0.4,
        random_state=12345))

We got a training sample = 60%, a validation sample - 40%. Now let's split the validation in half, and get the validation and test

In [20]:
features_valid, features_test, target_valid, target_test =(
    train_test_split(
        features_valid, target_valid, test_size=0.5,
        random_state=12345))

Let's look at the sizes of the resulting samples

In [21]:
for i in (features, features_train, features_test, features_valid, 
          target, target_train, target_test, target_valid):
    print(i.shape)

(10000, 11)
(6000, 11)
(2000, 11)
(2000, 11)
(10000,)
(6000,)
(2000,)
(2000,)


Everything as we wanted

### Using scaler

In [22]:
# Let's create a list with numeric columns
numeric = ['balance', 'estimated_salary', 'credit_score', 'age', 'tenure',]

In [23]:
scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])


In [24]:
features_train

credit_score       age    tenure   balance  num_of_products  \
7479     -0.886751 -0.373192  1.032538  1.232271                1   
3411      0.608663 -0.183385  1.032538  0.600563                1   
6027      2.052152  0.480939 -0.719791  1.027098                2   
1247     -1.457915 -1.417129  0.331606 -1.233163                2   
3716      0.130961 -1.132419 -1.070257  1.140475                1   
...            ...       ...       ...       ...              ...   
4478     -1.073677 -0.752805 -0.369325 -1.233163                2   
4094     -1.447531 -0.942612  1.733470 -1.233163                2   
3492      0.027113  0.575842  0.273665 -0.310229                1   
2177      0.151731 -1.417129 -0.369325 -1.233163                2   
4578      0.400966 -0.088482 -1.420723 -1.233163                2   

      has_cr_card  is_active_member  estimated_salary  geography_Germany  \
7479            1                 0         -0.187705                  0   
3411            0                 0         -0.333945                  0   
6027            0                 1          1.503095                  1   
1247            1                 0         -1.071061                  0   
3716            0                 0          1.524268                  1   
...           ...               ...               ...                ...   
4478            0                 1         -1.278361                  0   
4094            0                 1         -1.281307                  0   
3492            0                 1         -0.903158                  0   
2177            0                 1         -1.128539                  0   
4578            0                 0         -0.949841                  0   

      geography_Spain  gender_Male  
7479                1            1  
3411                0            0  
6027                0            1  
1247                0            1  
3716                0            0  
...               ...          ...  
4478                0            1  
4094                0            1  
3492                0            0  
2177                0            0  
4578                0            1  

[6000 rows x 11 columns]

## Problem research

We need to find out if the client leaves the bank or not, so we need to solve the classification problem, so only these models are suitable: decision tree, random forest and logistic regression

### Using a decision tree

In [25]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train, target_train) # train the model on the training set
    predictions_valid = model_decision.predict(features_valid) # get the predictions of the model on the validation выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 max:", f1_max, ' depth best:', depth_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc)


F1 max: 0.5696969696969697  depth best: 6
Roc_auc max: 0.7078133789824521  depth best: 6
CPU times: user 3.13 s, sys: 45.3 ms, total: 3.18 s
Wall time: 5.13 s


F1 = 0.569 depth best: 6, a little short of the required 0.59

### Let's try random forest

In the case of using RandomForestClassifier, when iterating over hyperparameters, we will iterate over n_estimators and max_depth

In [26]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 120, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train, target_train)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 max:", f1_max, ' depth best:', depth_f1, ' n_estimators best:', n_est_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc, ' n_estimators best:', n_est_roc_auc)

F1 max: 0.6026986506746627  depth best: 15  n_estimators best: 32
Roc_auc max: 0.7259359178315862  depth best: 15  n_estimators best: 22
CPU times: user 4min 28s, sys: 3.31 s, total: 4min 32s
Wall time: 5min 31s


F1 max: 0.605 depth best: 9 n_estimators best: 7 - slightly better than previous test

### And let's see the logistic regression

In [27]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
f1_score(target_valid, predicted_valid)

0.33389544688026984

F1 = 0.335. Let's try to get rid of the imbalance, and calculate f1 after that

## Fighting imbalance

### Upsampling

In [28]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_train_upsampled, target_train_upsampled = upsample(features_train, target_train, 4) #так как дисбаланс был в соотношении 4к1


### Check the decision tree after correcting the imbalance

In [29]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train_upsampled, target_train_upsampled) # train the model on the training set
    predictions_valid = model_decision.predict(features_valid) # get the predictions of the model on the validation выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
    if roc_auc_max < roc_auc_score(target_valid, predictions_valid):
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 max:", f1_max, ' depth best:', depth_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc)


F1 max: 0.5963791267305644  depth best: 5
Roc_auc max: 0.7587588238496482  depth best: 5
CPU times: user 2.56 s, sys: 22.2 ms, total: 2.59 s
Wall time: 2.57 s


F1 = 0.595, depth is the best: 5, we have already achieved the desired result

### Now random forest

In [30]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 60, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train_upsampled, target_train_upsampled)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 max:", f1_max, ' depth best:', depth_f1, ' n_estimators best:', n_est_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc, ' n_estimators best:', n_est_roc_auc)

F1 max: 0.6338912133891214  depth best: 9  n_estimators best: 37
Roc_auc max: 0.7881671193268771  depth best: 9  n_estimators best: 37
CPU times: user 1min 18s, sys: 612 ms, total: 1min 18s
Wall time: 1min 19s


F1 max: 0.631 depth best: 9 n_estimators best: 32

Remember these settings

As you can see, after upsampling with a random forest, we managed to achieve the required value of f1. Now let's try downsample

### Downsampling

In [31]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

features_train_downsampled, target_train_downsampled = downsample(features_train, target_train, 0.25)

### Let's check f1 in the decision tree

In [32]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train_downsampled, target_train_downsampled) # train the model on the training выборке
    predictions_valid = model_decision.predict(features_valid) #get the predictions of the model on the validation выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
    if roc_auc_max < roc_auc_score(target_valid, predictions_valid):
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 max:", f1_max, ' depth best:', depth_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc)


F1 max: 0.5943204868154158  depth best: 5
Roc_auc max: 0.7635631718072333  depth best: 5
CPU times: user 1.14 s, sys: 10.1 ms, total: 1.15 s
Wall time: 1.16 s


Got F1 max: 0.593 at depth = 5

### Now let's check the random forest

In [33]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 60, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train_downsampled, target_train_downsampled)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 max:", f1_max, ' depth best:', depth_f1, ' n_estimators best:', n_est_f1)
print('Roc_auc max:', roc_auc_max, ' depth best:', depth_roc_auc, ' n_estimators best:', n_est_roc_auc)

F1 max: 0.6061776061776061  depth best: 8  n_estimators best: 22
Roc_auc max: 0.7795610304925628  depth best: 5  n_estimators best: 22
CPU times: user 36.2 s, sys: 334 ms, total: 36.5 s
Wall time: 36.7 s


Got F1 max: 0.604 at depth = 8, n_estimators = 12

### And check the logistic regression

In [34]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train_downsampled, target_train_downsampled)
predicted_valid = model.predict(features_valid)
f1_score(target_valid, predicted_valid)

0.47965367965367967

Before dealing with the imbalance, we had the following maximum performance with these hyperparameters:

Decision tree - F1 max: 0.569 at depth = 6
Random forest - F1 max: 0.601 at depth = 9, n_estimators = 7
Logistic regression - F1 = 0.335

To combat the imbalance, we used the upsampling and downsampling methods.

After using the upsampling method, we got the following results:

Decision tree - F1 max: 0.596 at depth = 5
Random forest - F1 max: 0.63 at depth = 9, n_estimators = 47
Logistic regression - F1 = 0.49

After using the downsampling method, we got the following results:

Decision tree - F1 max: 0.597 at depth = 5
Random forest - F1 max: 0.605 at depth = 5, n_estimators = 37
Logistic regression - F1 = 0.485

As you can see, the best model is Random Forest after upsampling method: F1 maximum: 0.604 at depth = 8, n_estimators = 12

## Model testing

As we remember, we use the following data: n_estimators=42, max_depth=13 - this is our best model

In [35]:
model_random_forest = RandomForestClassifier(n_estimators=42, max_depth=13, random_state=12345)
model_random_forest.fit(features_train_upsampled, target_train_upsampled)
predicted_test = model_random_forest.predict(features_test)
f1 = f1_score(target_test, predicted_test)
roc_auc = roc_auc_score(target_test, predicted_test)

print("F1:", f1)
print('Roc_auc:', roc_auc)

F1: 0.6055488540410132
Roc_auc: 0.7475463631307612


F1 = 0.601 after upsampling, with n_estimators=42, max_depth=13

### Conclusion.
We carried out a primary analysis of the data, revealed a clear imbalance of classes, negative to positive - 4 to 1.

We pre-processed the data, filled in the gaps, made one_hot_encoding.

Before dealing with the imbalance, we had the following maximum performance with these hyperparameters:

**Decision tree** - F1 max: 0.569 at depth = 6

**Random forest** - F1 max: 0.605 at depth = 19, n_estimators = 47

**Logistic regression** - F1 = 0.335

To combat the imbalance, we used the upsampling and downsampling methods.

After using the upsampling method, we got the following results:

**Decision tree** - F1 max: 0.596 at depth = 5

**Random forest** - F1 max: 0.63 at depth = 9, n_estimators = 47

**Logistic regression** - F1 = 0.49

After using the downsampling method, we got the following results:

**Decision tree** - F1 max: 0.597 at depth = 5

**Random forest** - F1 max: 0.605 at depth = 5, n_estimators = 37

**Logistic regression** - F1 = 0.485


On testing, the models got f1 = 0.875 with n_estimators=47, max_depth=9